### PR TITLE
speedup singlelayer

### PR DESCRIPTION
**Pull request by lruthotto**
* made singleLayer more memory efficient and also faster (benchmark t…
* changed order of bias and batchnorm in singleLayer
* also fixed a bug in apply

### CODE REVIEW

**Comment by klensink** [comment]
Could you submit this to the dev branch?

File: src/layers/doubleSymLayer.jl [hunk 1]
```diff
@@ -39,21 +39,18 @@ end
...
-    tmp = Array{Any}(2)
     nex = div(length(Yin),nFeatIn(this))::Int
     Y   = reshape(Yin,:,nex)
 
     theta1,theta2,theta3,theta4 = splitWeights(this,theta)
     Kop    = getOp(this.K,theta1)
     KY     = Kop*Y
-    KY,dummy,tmp[1] = apply(this.nLayer,theta4,KY)
-    Yt     = KY
+	tmp    = (doDerivative) ? copy(KY) : (T)[]
```
Comment (DavidBegert) [inline src/layers/doubleSymLayer.jl]: Hi Lars, 
You now only store KY in tmp whereas before we stored KY as tmp[2] and had some other Ys stored in tmp[1] corresponding to the other Ys in batchNormNN.. is this all we need now? If so I can fix the conflicts and merge yours in.

Comment (lruthotto) [reply src/layers/doubleSymLayer.jl]: Hi David, there is no need anymore to store Ys. Note that doing this doubles the memory. Also, from KY I can re-compute everything thats needed without convolutions.  It would be great if you could help with the conflicts! Thanks

Comment (DavidBegert) [reply src/layers/doubleSymLayer.jl]: Awesome. That will help a lot with the memory . Thanks!

Comment (lruthotto) [reply src/layers/doubleSymLayer.jl]: Actually, while we're at it: More memory can be saved by writing a specialized bach norm layer (see MATLAB here: https://github.com/XtractOpen/Meganet.m/commit/c40b9a74cab1c59a5353958ad108e6d0e39c6e3a) I know you've recently combined them in batchNorm.jl but there you keep some tmp variables, right? Not needed for batch norm, I think!

**Review by DavidBegert — APPROVED** [review]
Pretty sure things work now. Good work!

File: src/layers/doubleSymLayer.jl [hunk 2]
```diff
@@ -85,47 +82,73 @@ function initTheta(this::DoubleSymLayer{T})  where {T<:Number}
...
+function JYmv(this::DoubleSymLayer{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp::Array{T})  where {T<:Number}
 
     nex = div(length(dY),nFeatIn(this))
     dY  = reshape(dY,:,nex)
     Y   = reshape(Y,:,nex)
     th1, th2,th3,th4  = splitWeights(this,theta)
 
+	# re-compute derivative of activation
+	KY              = copy(tmp);
+    KY,dummy,tmpNL  = apply(this.nLayer,th4,KY,true)
```
Comment (DavidBegert) [inline src/layers/doubleSymLayer.jl]: After merging the code in, the memory for an epoch increases about 2 GiB and gc time increases by 5% as well. Number of allocations decreases but I'm thinking it stems from this line since apply(nLayer) with doDerivatives=true allocates 3 copies of Y. what do we need tmpNL to be for this to work? How are we supposed to remove all tmps from batchNormLayer when it seems like we need it here?

File: src/layers/doubleSymLayer.jl [hunk 3]
```diff
@@ -85,47 +82,73 @@ function initTheta(this::DoubleSymLayer{T})  where {T<:Number}
...
+function Jmv(this::DoubleSymLayer{T},dtheta::Array{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp::Array{T})  where {T<:Number}
     nex = div(length(Y),nFeatIn(this))
 
     th1, th2,th3,th4    = splitWeights(this,theta)
     dth1,dth2,dth3,dth4 = splitWeights(this,dtheta)
 
-    Kop    = getOp(this.K,th1)
+	# re-compute derivative of activation
+	KY              = copy(tmp);
+    KY,dummy,tmpNL  = apply(this.nLayer,th4,KY,true)
```
Comment (DavidBegert) [inline src/layers/doubleSymLayer.jl]: And we are doing it a few times.. so instead of just doing it once and storing them all we are doing it for each function!

**Comment by DavidBegert** [comment]
@lruthotto not sure if you can see my other comments, but what exactly do we need to store when running batch norm? In the backward pass functions we are running batchNormNN with doDerivatives=true which uses a lot of memory storing the 3 Ys at each point of batch norm (one at start, one after norm layer, one after affine scaling layer). Causing memory footprint to grow a lot.
